bencoder/src/bencode.rs

```diff
@@ -119,38 +119,33 @@ impl Bencode {
         while data[i] != b':' {
             i += 1;
         }
-        let length = &data[0..i];
-        let length = match String::from_utf8(length.to_vec()) {
-            Ok(s) => s,
-            Err(_) => return Err(BencodeError::InvalidBencodeString),
-        };
-        let length = match length.parse::<i64>() {
-            Ok(n) => n,
-            Err(_) => return Err(BencodeError::InvalidBencodeString),
-        };
-        let mut i = i + 1;
-        let mut string = Vec::new();
-        for _ in 0..length {
-            string.push(data[i]);
-            i += 1;
-        }
-        Ok((Bencode::BString(string), i))
+        let length =
+            std::str::from_utf8(&data[0..i]).map_err(|_| BencodeError::InvalidBencodeString)?;
+        let length: usize = length
+            .parse()
+            .map_err(|_| BencodeError::InvalidBencodeString)?;
+
+        i += 1;
+
+        let string: Vec<u8> = data.iter().skip(i).take(length).cloned().collect();
+        let len = string.len();
+
+        Ok((Bencode::BString(string), i + len))
```

Comment:
This looks to me like it could become simply:
```
let string = String::from_bytes(&data[i..i+length]);
```
It would require slight changes to `Bencode::BString` and I don't 100% recall the name of the constructor function but this illustrates the idea.

Reply:
This would be wrong because BString takes a `Vec<u8>`, this is because "strings" are "byte strings" in bencode, and they allow non-utf8 data for strings not meant to be presented to the user.